feat: make yamlfix configurable with sane default values

marcules opened the pull request:
<!-- Describe what the change is, trying to link it with open issues -->
This fixes #117 and fixes #190 

- make default values, like line_length, explicit_start and indentation configurable
- add configuration option for none-value representation
- add configuration option for quote-style representation
- add configuration option for flow-style / block-style sequences
- add an option to quote every key and value
- add an option to quote only simple values
- add an option to switch between multi-line flow-style and flow-style
  -> block-style conversion if the list doesn't fit the line_length
- implement reading of configuration values from files (with maison)
- implement overriding specific values with prefixed env vars

## Checklist

* [x] Add test cases to all the changes you introduce
* [x] Update the documentation for the changes

marcules commented on .pre-commit-config.yaml:
```diff
@@ -1,27 +1,29 @@
-  - repo: https://github.com/flakehell/flakehell/
-    rev: v.0.9.0
+        additional_dependencies:
+          - pydantic
```
I had to add the additional dependency "pydantic" here, because it was not present on my machine, but was necessary to run the pre-commit-hook.

See also the discussion here: https://gitter.im/tiangolo/fastapi?at=5dfcda8844e1fb33f6fc13d2

marcules commented on .pre-commit-config.yaml:
```diff
@@ -1,27 +1,29 @@
-      - name: Run flakehell static analysis tool
-        id: flakehell
+      - name: Run flakeheaven static analysis tool
+        id: flakeheaven
```
I replaced flakehell with flakeheaven here, as it seems like flakehell is not maintained well, and also flakeheaven was already used in pyproject.toml

lyz-code commented on src/yamlfix/entrypoints/cli.py:
```diff
@@ -18,21 +18,62 @@ def _format_file_list(files: Tuple[TextIOWrapper]) -> str:
-@click.command()
-@click.version_option(version="", message=version.version_info())
-@click.option("--verbose", is_flag=True, help="Enable verbose logging.")
+@click.command()  # type: ignore
```
why did you add all these type ignore? they were not raising any issue before

lyz-code commented on src/yamlfix/services.py:
```diff
@@ -128,6 +141,23 @@ def fix_code(source_code: str) -> str:
     else:
         shebang = ""
 
+    configurers = [
```
I'd extract all the configuration of the yaml formatter to an internal function to keep `fix_code` simpler to read

lyz-code replied:
Or maybe create a new class `Yaml` in `adapters.py` that acts as interface over ruyaml hiding the details to this service.

lyz-code commented on src/yamlfix/services.py:
```diff
@@ -139,30 +169,150 @@ def fix_code(source_code: str) -> str:
+    log.debug("Running ruamel yaml 'none representation' configuration...")
+
+    def represent_none(self, data):
+        # type: (Any, Any) -> Any
```
why didn't you add the typing information in the function definition?

lyz-code commented on src/yamlfix/services.py:
```diff
@@ -18,18 +18,29 @@
 Files = Union[Tuple[TextIOWrapper], List[str]]
 
 
+class RuamelOptions(object):
```
Maybe this file belongs better in a `model.py` file. I hadn't created it as so far we didn't have any models

lyz-code commented on src/yamlfix/entrypoints/cli.py:
```diff
@@ -18,21 +18,62 @@ def _format_file_list(files: Tuple[TextIOWrapper]) -> str:
+@click.option(
+    "--quote-everything", is_flag=True, help="Quote all keys and simple string values."
+)  # type: ignore
+@click.option(
```
So far I don't understand very well the difference between `--quote-everything` and `--quote-basic-values`. I'm sure that with the tests it will be clearer. Just a note to keep in mind when writing the docs of these features

lyz-code commented on src/yamlfix/services.py:
```diff
@@ -128,6 +141,23 @@ def fix_code(source_code: str) -> str:
+    yaml = ruyaml.main.YAML()
+
+    if ruamel_options is None:
+        # Always run base configuration, even if no ruamel_options are provided
```
If we're going to run despite the value of `ruamel_options`, shouldn't it be clearer to have the next snippet instead?:

```python
_ruamel_yaml_configure_base(...)
if ruamel_options is not None:
  for configurer in configuresrs:
     ...
```

lyz-code reviewed (changes requested):
Hi @marcules thank you so much for the complete PR! it's awesome.

Everything looks good so far, I'm missing the possibility of configuring `yamlfix` through a file [like we do in `autoimport`](https://github.com/lyz-code/autoimport/pull/160/files) though, but if you don't feel like implementing it, it's fine.

Can't wait to see this merged :)

marcules commented:
Hi @lyz-code - I refactored the code as requested and hope that I addressed everything you mentioned.

Regarding the topic of adding maison, I'm not really sure I can implement that. Currently it looks like flakeheaven cannot be used by flake8 >4 which makes me unable to add any new dependency with `pdm add maison` if I don't update all (or rather some unknown quantity of) dependencies.

```bash
in yamlfix on feature/make-null-value-configurable marcyamlfix
➜ pdm add maison
Adding packages to default dependencies: maison
See /var/folders/9k/0s245qrj13338vfl15y901zc0000gn/T/pdm-lock-4_r3ckaf.log for detailed debug log.
[InvalidSpecifier]: Invalid specifier: '>=3.5.*'
Add '-v' to see the detailed traceback
```

It looks like flakeheaven would need to be updated to not use the deprecated extra verbose flag from flake8 anymore: https://github.com/PyCQA/flake8/commit/929cf5dfd3cf269e4eff1b46dd28bd19d15bcdee

I'm still unsure about the approach regarding Any types - flake8 says I can't use them, but I don't know what type is being returned, and the original source contains only type comments with type Any.

Docs + Tests are still missing, but just wanted to make sure this is going in the right direction, I'll try implementing some tests and updating the docs in the coming days.

marcules commented:
I've figured out what the issue with my local setup was regarding pdm.
I'm using python 3.10.8 on a M1 mac - but some components in this project require python 3.7.

I was able to install python3.7 successfully now with https://stackoverflow.com/a/70327233
And afterwards I had to create my virtual environment specifically with that python version with https://stackoverflow.com/a/39713544

And after THAT I was able to install pdm within the venv and use that specific pdm version to install maison.
`➜ .venv/bin/pdm add maison`

I'll look into adding these configs with the config file (and I've also seen in the original issue, that list-style and line-length was requested as well, so I'll add that too).

No guarantees how fast though :)

lyz-code commented on pyproject.toml:
```diff
@@ -125,7 +126,6 @@ fixers = [
 ]
 typing = [
     "mypy>=0.910",
-
```
I think you added a space here by mistake

lyz-code reviewed (changes requested):
It is soooo beautiful I want to cry! `(╥﹏╥)`.

lyz-code commented:
> I'm still unsure about the approach regarding Any types

Don't worry about it, if you can avoid using `Any` specify a better type, if you
can't then ad a `type: ignore` or whatever you need.

> Docs + Tests are still missing, but just wanted to make sure this is going in the right direction,

You definitely are! It was a joy to read the code

> sorry for the many failing tests,

Don't worry! I don't use very much the pre-commit hooks so they may not be in
sync with the CI :(

> I've figured out what the issue with my local setup was regarding pdm.

Damn I'm sorry it's been difficult to debug this, and thank you for taking the
efforts to debug so deep the problem.

> I'll look into adding these configs with the config file 

Awesome!

> No guarantees how fast though :)

The issue has been open for months, no rush at all

marcules commented:
Hi @lyz-code I just finished implementing the maison-configuration parser and I added a environment-parser as well, to be able to override specific config with `YAMLFIX_`-env vars instead of / in addition to providing the file, for easier CI "one off" configs - so users do not have to maintain a separate config file if they don't want to.

I have also implemented the possibility to "force" a list-style -> there are two styles which ruyaml calls "flow-style" and "block-style", this can now be toggled with the config `flow_style_sequence=true` for flow-style and `flow_style_sequence=false` for block style. I have also implemented the discussion topic in the original issue, where there was a discussion between multiline-flow-style sequences not being desired, but rather the desired behaviour being "if it doesn't fit in the line, it should use block-style instead" -> so I have implemented that as well with the `flow_style_sequence_multiline=true` flag, which allows, flow-style on multiple lines if true, and forces block-style if the line is too long otherwise.

I've also guarded those two options behind `flow_style_sequence_enabled`, as there is one test, that would now be converted to block-style automatically otherwise and set this flag to `flow_style_sequence_enabled=false` as a default (no "list shenanigans" per default 🤣)

I know the changes are now rather big, but it is also implementing a bunch of configurability - please let me know if I should change something to match your / the projects style.

One last thing - I'll try to write some tests this week as well, but no guarantees. In particular the list behaviour needs some tests, as I have implemented special handling for lists that contain comments, or lists that contain non-scalar values (i.e. anything that's not a simple string, int, bool).

lyz-code commented on src/yamlfix/adapters.py:
```diff
@@ -104,6 +110,147 @@ def _configure_quotation_for_basic_values(self) -> None:
+
+    def _configure_flow_style_for_sequences(self) -> None:
+        """Configure Ruamel's yaml fixer to represent lists as either block-style \
+            or flow-style.
```
I'm not familiar with the differences between `block-style` and `flow-style`. Other code readers may not be either, could you explain the differences in this docstring (and in the doc for the users too)

lyz-code commented on src/yamlfix/adapters.py:
```diff
@@ -104,6 +110,147 @@ def _configure_quotation_for_basic_values(self) -> None:
+            if isinstance(seq_node.comment, list):
+                for comment in seq_node.comment:
+                    if isinstance(comment, list):
+                        comment_tokens = comment_tokens + comment
```
maybe use `comments_tokens.extend(comment)` instead

lyz-code commented on src/yamlfix/adapters.py:
```diff
@@ -104,6 +110,147 @@ def _configure_quotation_for_basic_values(self) -> None:
+                ):
+                    comment_tokens = comment_tokens + node.comment
+
+            return any(
```
Beautiful return construction

lyz-code commented on src/yamlfix/adapters.py:
```diff
@@ -104,6 +110,147 @@ def _configure_quotation_for_basic_values(self) -> None:
+
+            # comma and space between scalar values or colon and space
+            # between key + values
+            seperator_length: int = 2
```
possible typo `seperator` -> `separator`

lyz-code commented:
Damn @marcules I'm sorry, I should have realized that `maison` might not be the best configuration management for this case. Please read the comments I made on the issue. 

I'm not sure how to proceed on the configuration through file. On one side, `goodconf` supports loading from environmental variables by default, and having a configuration in yaml for a yaml tool makes more sense as it's more pleasant to read than the toml/ini format. However the `maison` code is interesting for the Python projects because you avoid polluting the root of your project with configuration files. As I said in the issue maybe we can have both, but that adds complexity... After reviewing all the PR, I'd say we stay with `maison` and don't use `goodconf`.

Regarding the code, I'm afraid that I don't fully understand what you're doing in some of the functions because I'm not familiar with `ruamel` internals. Therefore I'd need you to do a similar awesome job as you're doing with the code, but with the tests to be able to merge this. A full coverage of the new code will let me maintain the package even if today I don't understand everything if ever bugs appear. 

And yet again you almost made me cry. The new code is as beautiful as the previous one. I'm amazed with your control of `ruamel` internals. I've always been very lazy to dive into their code and did the fixes on their returned string. Your approach is much more elegant. Probably some of the existent fixes could be reimplemented with your method, making the tool faster.

Seeing that you're very interested in the program and know better than I the internals of `ruamel`, I wanted to propose you to help me in the maintenance of the program. What do you think?

lyz-code commented:
I've just thought that maybe it's interesting to allow the user to override a configuration from the file itself. I'm thinking of https://github.com/lyz-code/yamlfix/issues/191 . I know that I would not want that option enabled by default, but I will want it in one specific file. I could not add a `yamlfix.ini` in that directory because it lives with other files where I don't want the feature.

We could implement special comments (like vim does) at the end of the document that override the rest of the configurations.

If you don't feel like implementing this in this PR I'll create a new issue for it

marcules commented:
> We could implement special comments (like vim does) at the end of the document that override the rest of the configurations.

Please create a new issue for it, I think it would make this PR even more complex if we add it here too. Also I'm not sure how much time I can invest in the PR, so I'm trying to finish the tests so this topic can be "closed" :)

Regarding that configuration as annotation, in the link I proposed a whitelist/blacklist for matching specific keys within a file. But we could also extend that construct to be something like:

`overrides.{fileNamePattern}.keyPath.{keyPath}.{config_key}: {config_value}`
and
`overrides.{fileNamePattern}.recursiveKeyPath.{keyName}.{config_key}: {config_value}`

and then apply whatever is in config_key/config_value for that specific node (or nodes).
Not sure if that would make it better/easier to configure, but at least the user does not have to modify the files to add metadata to it, and also with patterns in both filename+key the user could configure a set of settings for multiple similar files.

Additionally this could then be just passed as an environment variable `YAMLFIX_overrides.values_yaml.keyPath.global_key_subkey.flow_style_sequence=true` or something like that (with special handling for underscore -> subpath/glob conversion).

Thinking about it, you approach seems more straight forward, indeed.

lyz-code commented:
Done https://github.com/lyz-code/yamlfix/issues/192

marcules commented:
@lyz-code I'm not sure why the tests are now randomly failing, I suppose it has to do with my two new e2e tests.
It looks like the YamlfixConfig class is somehow cached/re-used between tests? Not sure...

marcules commented:
> If you mean the tests of [your last commit build](https://github.com/lyz-code/yamlfix/actions/runs/3694879517/jobs/6256592220) it looks like the changes in the PR have changed the default representation of `null` to `~`.

I figured out what the issue was. When I was monkey-patching the representation class, the class itself could be reused in another test. I have now refactored it to use proper inheritance (the new class extends RoundTripRepresenter) and now it is stable and working as intended.

Just wrapping up the tests now.

marcules commented:
Btw. @lyz-code I'm not sure if you're aware, but from the maintainer of ruyaml there's also a "plugin" that supports jinja yaml files. I've seen there were some fixes for that in the code. I think that could be refactored / replaced with that plugin.

https://pypi.org/project/ruamel.yaml.jinja2/

marcules commented:
Hi @lyz-code I'm finished with the tests now and I've also rebased and tidied up the commit history for this branch and updated the initial title + description.

I hope everything is to your satisfaction and that we can get this merged 🚀 

I think the only things that are missing now are "docs" and "version bump" < I'm not sure on how to do both of those, would it be OK if you do that?

lyz-code commented on tests/unit/test_adapter_yaml.py:
```diff
@@ -0,0 +1,567 @@
+              complex_key2: {quote}value{quote}
+              list:
+                - item1
+                - item2
```
If we're quoting strings I'd also expect to have quoted strings in lists. In what case do you think it's interesting not to have quoted the strings in lists?

lyz-code commented on tests/unit/test_adapter_yaml.py:
```diff
@@ -0,0 +1,567 @@
+            list: [loooooooooooooooooooongItem, loooooooooooooooooooongItem,
+              loooooooooooooooooooongItem, loooooooooooooooooooongItem,
+              loooooooooooooooooooongItem, loooooooooooooooooooongItem]
+            list2: [item, item, item]
```
I don't think this way of writing long lists is good. I'd remove this functionality. In what cases do you think it's interesting?

lyz-code commented on tests/unit/test_adapter_yaml.py:
```diff
@@ -0,0 +1,567 @@
+
+            key: value
+            """
+        )
```
In this case I'd expect the output to be:

```yaml
---
list: [item, item]

key: value
```

Or 

```yaml
---
list: [item, item]
key: value
```

Having many whitelines between elements doesn't look like a good idea. In which cases do you think they're interesting?

lyz-code replied:
Recorded in a new issue https://github.com/lyz-code/yamlfix/issues/193

lyz-code commented on src/yamlfix/model.py:
```diff
@@ -0,0 +1,23 @@
+class YamlfixConfig(ConfigSchema):
+    """Configuration entity for yamlfix."""
+
+    allow_duplicate_keys: bool = True
```
I'd say we use `allow_duplicate_keys: bool = False` as default, what do you think?

marcules replied:
@lyz-code I set this to `False` and enabled it for the one test, that tested on merge keys.

lyz-code commented on src/yamlfix/model.py:
```diff
@@ -0,0 +1,23 @@
+    config_path: Optional[str] = None
+    explicit_start: bool = True
+    flow_style_sequence_multiline: bool = False
+    flow_style_sequence: Optional[bool] = None
```
I'd say we use `flow_style_sequence: bool = True` as default, what do you think?

lyz-code reviewed (changes requested):
Hi @marcules, the tests are very clear, thank you. There are some functionality I'm not sure I want to have and maintain unless you think there are real cases where it's interesting. 

Please review the comments I've made and answer on each of them to make it easier for me to review what you mean on each case.

I'll handle the bumping of the version once this is merged. However I won't have time to write the docs and I feel you know better your code so your explanation would be better than mine. I can guide you through it though. You can reuse [the docs of autoimport](https://lyz-code.github.io/autoimport/#configuration) and tweak to our case. I think it's important to do a description of what each flag does with a little example of how it will transform the text.

Thanks again for your time, can't wait to have this merged :)

marcules commented:
Hi @lyz-code I hope everything is in order now, I have implemented the list-quotation like requested and also updated the docs. My vacation is over now, so if something major is missing I probably won't be able to really get into it in any acceptable timeframe.

Hope it looks good now though and that it can be merged! 👍

lyz-code commented on docs/index.md:
````diff
@@ -76,6 +76,287 @@ yamlfix will do the following changes in your code:
+export YAMLFIX_ALLOW_DUPLICATE_KEYS="true"
+```
+
+This option enables the [ruyaml duplicate keys check](https://ruyaml.readthedocs.io/en/latest/api.html#duplicate-keys). With this check enabled `yamlfix` will throw an error if the same key is defined more than once in a dict. Before enabling this option, you should be aware, that multiple yaml-anchor merge keys are considered duplicate keys and should be provided as a list instead: https://githu ...
````
This paragraph made me confused. If I set `allow_duplicate_keys` I expect to be able to have duplicate keys in the yaml, but from what I read here, if I set it to `True` it will enable the duplicate check that will raise an exception if there are duplicate keys. 

If I misunderstood the paragraph then I'll ask you to reformulate it to prevent other people of getting the same idea.

If I understood it right then I think that the exception should be raised if `allow_duplicate_keys == False` instead. Keeping the default to `True`

marcules replied:
Hi @lyz-code I think I got confused while writing that (double negatives), I'll re-phrase it.

lyz-code reviewed (changes requested):
Awesome job as always @marcules. Sad to hear that your holidays are over, I hope that the time spent on developing this PR was at least pleasant.

One minor last change may be needed, it shouldn't require too much coding to solve. It has to do with the `allow_duplicate_keys` configuration, please read the two remaining unsolved comments

And tnanks again for all your hard work

marcules commented:
Hey @lyz-code sorry, I just realized now, that I didn't ever "commit" my comments / answers to your questions until now... OMG :D

lyz-code reviewed (approved):
Niiiiiice it's ready Yay!!

lyz-code commented:
Available on `1.2.0`
